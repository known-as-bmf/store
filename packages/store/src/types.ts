import {
  HookSubscriber,
  TransformInput,
  TapInput,
  TapOutput,
} from '@known-as-bmf/hookable';

///@ts-ignore
export interface Store<S> {}

export interface Hooks<S> {
  /**
   * Register a function that will be invoked each time a state change is requested.
   * This function can transform the state and must return an array containing the new state.
   * An array must be returned because of implementation details behind the scene.
   * @param fn A state transformation function.
   */
  transformState: HookSubscriber<TransformInput<(state: S) => S>>;
  /**
   * Register a function that will be invoked when the state is about to change.
   * It is invoked after all `transformState` hooks.
   * @param fn A function invoked when the state is about to change.
   */
  stateWillChange: HookSubscriber<TapInput<(state: S) => S>>;
  /**
   * Register a function that will be invoked after the state changed.
   * It is invoked after all `transformState` and `stateWillChange` hooks.
   * @param fn A function invoked when the state has changed.
   */
  stateDidChange: HookSubscriber<TapOutput<(state: S) => S>>;
}

export interface StateChangedEvent<S> {
  /**
   * The previous value of the state.
   */
  previous: S;
  /**
   * The new value of the state.
   */
  current: S;
}

export type Selector<S, R> = (state: S) => R;
export type SubscriptionCallback<S> = (event: StateChangedEvent<S>) => void;

/**
 * Middleware function signature.
 */
export type Middleware<S> = (store: Store<S>, hooks: Hooks<S>) => void;
