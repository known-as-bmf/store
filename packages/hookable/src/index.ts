export { createHookable } from './hookable';
export { every, once } from './hook-subscribers';
export {
  Hook,
  HookSubscriber,
  Hooks,
  TapInput,
  TapOutput,
  TransformInput,
  TransformOutput,
} from './types';
