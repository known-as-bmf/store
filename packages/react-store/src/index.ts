export { useStore } from './useStore';
export { useNewStore } from './useNewStore';
