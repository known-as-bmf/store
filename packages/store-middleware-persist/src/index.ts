export { PersistOptions } from './types';
export { persistMiddleware } from './persist';
