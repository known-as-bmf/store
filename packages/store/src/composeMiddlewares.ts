import { Middleware } from './types';

/**
 * Compose middlewares from right to left.
 * @param middlewares Middlewares to compose.
 */
export const pipeMiddlewares = <S>(
  ...middlewares: Middleware<S>[]
): Middleware<S> => (store, hooks) => middlewares.map((m) => m(store, hooks));

/**
 * Compose middlewares from left to right.
 * @param middlewares Middlewares to compose.
 */
export const composeMiddlewares = <S>(
  ...middlewares: Middleware<S>[]
): Middleware<S> => {
  return pipeMiddlewares(...middlewares.reverse());
};
