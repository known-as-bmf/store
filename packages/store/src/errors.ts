export const errors = {
  invalidStore: 'Provided value is not a valid store instance.',
};
