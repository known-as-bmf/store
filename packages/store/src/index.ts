export {
  Hooks,
  Middleware,
  Selector,
  StateChangedEvent,
  Store,
  SubscriptionCallback,
} from './types';
export { deref, of, set, subscribe, swap } from './store';
export { composeMiddlewares, pipeMiddlewares } from './composeMiddlewares';
