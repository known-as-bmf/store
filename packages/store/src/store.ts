import { produce } from 'immer';
import { createHookable } from '@known-as-bmf/hookable';

import { errors } from './errors';
import {
  SubscriptionCallback,
  Selector,
  StateChangedEvent,
  Hooks,
  Store,
  Middleware,
} from './types';

// Internal symbols used by the store instance.
/**
 * Current subscriptions to store changes.
 */
const SUBSCRIPTIONS = Symbol();
/**
 * Current state of the store.
 */
const STATE = Symbol();
/**
 * State transformation (middleware).
 */
const SET_STATE = Symbol();

/**
 * The store class.
 */
class StoreImpl<S> implements Store<S> {
  public declare [SUBSCRIPTIONS]: Set<Subscription<S, unknown>>;
  public declare [STATE]: S;
  public declare [SET_STATE]: (state: S) => S;
}

/**
 * Internal subscription type.
 */
type Subscription<S, R> = [Selector<S, R>, SubscriptionCallback<S>];

/**
 * Check if a value is a Store instance.
 * @param input The value to check.
 */
function isStore<S = unknown>(input: unknown): input is StoreImpl<S> {
  return input instanceof StoreImpl;
}

function assertStore<S>(store: Store<S>): asserts store is StoreImpl<S> {
  if (!isStore(store)) {
    throw new TypeError(errors.invalidStore);
  }
}

function getState<S>(store: StoreImpl<S>): S {
  const { [STATE]: state } = store;
  return state;
}

function setState<S>(store: StoreImpl<S>, state: S): void | never {
  store[SET_STATE](state);
}

/**
 * Notify all subscribers to the store that the state has changed.
 * If a selector was provises at subscription,
 * only notify if the desired value has changed.
 * @param store The store instance.
 * @param event The change event.
 */
function notifySubscribers<S>(
  store: StoreImpl<S>,
  event: StateChangedEvent<S>
): void {
  store[SUBSCRIPTIONS].forEach(([selector, callback]) => {
    if (selector(event.previous) !== selector(event.current)) {
      callback(event);
    }
  });
}

/**
 * Creates a store.
 * @param initialState The initial value of the state.
 * @param middleware Middleware to use for this store. You can compose multiple
 * middlewares with `composeMiddlewares` and `pipeMiddlewares`.
 */
export function of<S>(initialState: S, middleware?: Middleware<S>): Store<S> {
  const store: StoreImpl<S> = Object.defineProperties(
    Object.create(StoreImpl.prototype),
    {
      [SUBSCRIPTIONS]: {
        value: new Set<Subscription<S, unknown>>(),
        writable: true,
      },
      [STATE]: { value: undefined, writable: true },
    }
  ) as StoreImpl<S>;

  const setStateHook = createHookable(
    (s: S): S => {
      store[STATE] = s;
      return s;
    }
  );

  Object.defineProperty(store, SET_STATE, { value: setStateHook });

  const hooks: Hooks<S> = {
    stateDidChange: setStateHook.leave,
    stateWillChange: setStateHook.enter,
    transformState: setStateHook.transformInput,
  };

  if (middleware) {
    middleware(store, hooks);
  }

  setState(store, initialState);

  return store;
}

/**
 * Returns the current state of a store.
 * @param store The store you want to get the current state from.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function deref<S>(store: Store<S>): S {
  assertStore(store);

  return getState(store);
}

/**
 * Changes the state of a store using a function.
 * @param store The store of which you want to change the state.
 * @param mutationFn The function used to compute the value of the future state.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 * @throws {Error} if the new state does not pass validation.
 */
export function swap<S>(store: Store<S>, mutationFn: (current: S) => S): void {
  assertStore(store);

  const previous = getState(store);
  const current = produce(previous, mutationFn) as S;

  setState(store, current);
  notifySubscribers(store, { previous, current });
}

/**
 * Changes the state of a store with a new one.
 * @param store The store of which you want to change the state.
 * @param current The new state.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 * @throws {Error} if the new state does not pass validation.
 */
export function set<S>(store: Store<S>, current: S): void {
  assertStore(store);

  const previous = getState(store);

  setState(store, current);
  notifySubscribers(store, { previous, current });
}

/**
 * Subscribes to state changes.
 * @param store The store you want to subscribe to.
 * @param callback The function to call when the state changes.
 * @returns An unsubscribe function for this specific subscription.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function subscribe<S>(
  store: Store<S>,
  callback: SubscriptionCallback<S>
): () => void;
/**
 * Subscribes to state changes.
 * @param store The store you want to subscribe to.
 * @param callback The function to call when the state changes.
 * @param selector The selector function, narrowing down the part of the state you want to subscribe to.
 * @returns An unsubscribe function for this specific subscription.
 * @throws {TypeError} if the store is not a correct `Store` instance.
 */
export function subscribe<S, R>(
  store: Store<S>,
  callback: SubscriptionCallback<S>,
  selector: Selector<S, R>
): () => void;
export function subscribe<S>(
  store: Store<S>,
  callback: SubscriptionCallback<S>,
  selector: Selector<S, unknown> = (s): S => s
): () => void {
  assertStore(store);

  const subscription: Subscription<S, unknown> = [selector, callback];

  store[SUBSCRIPTIONS].add(subscription);
  return () => {
    store[SUBSCRIPTIONS].delete(subscription);
  };
}
