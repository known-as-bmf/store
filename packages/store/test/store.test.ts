import { of, deref, set, subscribe, swap, Store, Hooks } from '../src';

describe('of', () => {
  it('should be a function', () => {
    expect(of).toBeDefined();
    expect(of).toBeInstanceOf(Function);
  });

  it('should return something', () => {
    expect(of(1)).toBeDefined();
  });

  it('should return an object', () => {
    expect(typeof of(1)).toBe('object');
  });

  it('should set the initial state', () => {
    expect(deref(of(1))).toBe(1);
  });

  it('should have no enumerable properties', () => {
    expect(Object.keys(of(1)).length).toBe(0);
  });

  it('should properly set up middleware', () => {
    const transformState = jest.fn<[string], [string]>((s) => [s]);
    const transformStateSubscriber = jest.fn<(str: string) => [string], []>(
      () => transformState
    );

    const stateWillChange = jest.fn();
    const stateWillChangeSubscriber = jest.fn<(str: string) => void, []>(
      () => stateWillChange
    );

    const stateDidChange = jest.fn();
    const stateDidChangeSubscriber = jest.fn<(str: string) => void, []>(
      () => stateDidChange
    );

    const middleware = jest.fn<void, [Store<string>, Hooks<string>]>(
      (_, hooks) => {
        hooks.transformState(transformStateSubscriber);
        hooks.stateWillChange(stateWillChangeSubscriber);
        hooks.stateDidChange(stateDidChangeSubscriber);
      }
    );
    const store = of<string>('a', middleware);

    expect(middleware).toHaveBeenCalledTimes(1);
    expect(middleware).toHaveBeenCalledWith(store, expect.any(Object));

    expect(transformStateSubscriber).toHaveBeenCalledTimes(1);
    expect(transformStateSubscriber).toHaveBeenLastCalledWith(
      expect.any(Function)
    );
    expect(stateWillChangeSubscriber).toHaveBeenCalledTimes(1);
    expect(stateWillChangeSubscriber).toHaveBeenLastCalledWith(
      expect.any(Function)
    );
    expect(stateDidChangeSubscriber).toHaveBeenCalledTimes(1);
    expect(stateDidChangeSubscriber).toHaveBeenLastCalledWith(
      expect.any(Function)
    );

    expect(transformState).toHaveBeenCalledTimes(1);
    expect(transformState).toHaveBeenLastCalledWith('a');
    expect(stateWillChange).toHaveBeenCalledTimes(1);
    expect(stateWillChange).toHaveBeenLastCalledWith('a');
    expect(stateDidChange).toHaveBeenCalledTimes(1);
    expect(stateDidChange).toHaveBeenLastCalledWith('a');

    set(store, 'b');

    // no further call of middleware / subscribers
    expect(middleware).toHaveBeenCalledTimes(1);
    expect(transformStateSubscriber).toHaveBeenCalledTimes(1);
    expect(stateWillChangeSubscriber).toHaveBeenCalledTimes(1);
    expect(stateDidChangeSubscriber).toHaveBeenCalledTimes(1);

    expect(transformState).toHaveBeenCalledTimes(2);
    expect(transformState).toHaveBeenLastCalledWith('b');
    expect(stateWillChange).toHaveBeenCalledTimes(2);
    expect(stateWillChange).toHaveBeenLastCalledWith('b');
    expect(stateDidChange).toHaveBeenCalledTimes(2);
    expect(stateDidChange).toHaveBeenLastCalledWith('b');
  });
});

describe('deref', () => {
  it('should be a function', () => {
    expect(deref).toBeDefined();
    expect(deref).toBeInstanceOf(Function);
  });

  it('should return the latest state', () => {
    const store = of(1);

    expect(deref(store)).toBe(1);

    set(store, 2);

    expect(deref(store)).toBe(2);
  });

  it('should throw if not given a store', () => {
    expect(() => deref({})).toThrowErrorMatchingSnapshot();
    expect(() => deref({ ...of(1) })).toThrowErrorMatchingSnapshot();
  });
});

describe('swap', () => {
  it('should be a function', () => {
    expect(swap).toBeDefined();
    expect(swap).toBeInstanceOf(Function);
  });

  it('should change the latest state', () => {
    const store = of(1);

    expect(deref(store)).toBe(1);

    swap(store, (s) => s + 1);

    expect(deref(store)).toBe(2);

    swap(store, () => 0);

    expect(deref(store)).toBe(0);
  });

  it('should throw if not given a store', () => {
    expect(() => swap({}, jest.fn())).toThrowErrorMatchingSnapshot();
    expect(() => swap({ ...of(1) }, jest.fn())).toThrowErrorMatchingSnapshot();
  });
});

describe('set', () => {
  it('should be a function', () => {
    expect(set).toBeDefined();
    expect(set).toBeInstanceOf(Function);
  });

  it('should change the latest state', () => {
    const store = of(1);

    expect(deref(store)).toBe(1);

    set(store, 2);

    expect(deref(store)).toBe(2);
  });

  it('should throw if not given a store', () => {
    expect(() => set({}, 0)).toThrowErrorMatchingSnapshot();
    expect(() => set({ ...of(1) }, 0)).toThrowErrorMatchingSnapshot();
  });
});

describe('subscribe', () => {
  it('should be a function', () => {
    expect(subscribe).toBeDefined();
    expect(subscribe).toBeInstanceOf(Function);
  });

  describe('without selector', () => {
    it('should call the subscriber', () => {
      const initialState = { a: 1, b: 10 };
      const nextState = { a: 1, b: 10 };
      const callback = jest.fn();
      const store = of(initialState);

      subscribe(store, callback);

      set(store, nextState);

      expect(callback).toHaveBeenCalledTimes(1);
      expect(callback).toHaveBeenCalledWith({
        previous: initialState,
        current: nextState,
      });
    });
  });

  describe('with selector', () => {
    it('should call the subscriber if selector(state) changes', () => {
      const initialState = { a: 1, b: 10 };
      const nextState = { a: 2, b: 10 };
      const callback = jest.fn();
      const store = of(initialState);

      subscribe(store, callback, (s) => s.a);

      set(store, nextState);

      expect(callback).toHaveBeenCalledTimes(1);
      expect(callback).toHaveBeenCalledWith({
        previous: initialState,
        current: nextState,
      });
    });

    it("should not call the subscriber if selector(state) doesn't change", () => {
      const initialState = { a: 1, b: 10 };
      const callback = jest.fn();
      const store = of(initialState);

      subscribe(store, callback, (s) => s.a);

      swap(store, (s) => {
        s.b = 10;
        return s;
      });

      expect(callback).not.toHaveBeenCalled();
    });
  });

  it('should return a function', () => {
    const initialState = { a: 1, b: 10 };
    const callback = jest.fn();
    const store = of(initialState);

    const unsubscribe = subscribe(store, callback);

    expect(unsubscribe).toBeDefined();
    expect(unsubscribe).toBeInstanceOf(Function);
  });

  it('should unsubscribe when the returned function is called', () => {
    const initialState = { a: 1, b: 10 };
    const nextState = { a: 1, b: 10 };
    const callback = jest.fn();
    const store = of(initialState);

    const unsubscribe = subscribe(store, callback);

    set(store, nextState);

    unsubscribe();

    set(store, nextState);

    expect(callback).toHaveBeenCalledTimes(1);
  });

  it('should throw if not given a store', () => {
    expect(() => subscribe({}, jest.fn())).toThrowErrorMatchingSnapshot();
    expect(() =>
      subscribe({ ...of(1) }, jest.fn())
    ).toThrowErrorMatchingSnapshot();
  });
});
