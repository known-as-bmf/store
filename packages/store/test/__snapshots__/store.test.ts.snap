// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deref should throw if not given a store 1`] = `"Provided value is not a valid store instance."`;

exports[`deref should throw if not given a store 2`] = `"Provided value is not a valid store instance."`;

exports[`set should throw if not given a store 1`] = `"Provided value is not a valid store instance."`;

exports[`set should throw if not given a store 2`] = `"Provided value is not a valid store instance."`;

exports[`subscribe should throw if not given a store 1`] = `"Provided value is not a valid store instance."`;

exports[`subscribe should throw if not given a store 2`] = `"Provided value is not a valid store instance."`;

exports[`swap should throw if not given a store 1`] = `"Provided value is not a valid store instance."`;

exports[`swap should throw if not given a store 2`] = `"Provided value is not a valid store instance."`;
